### DISCUSSION TITLE
Resources for Peggy Grammar

### DISCUSSION THREAD

**Discussion by coffeeDev98** [mliebelt/pgn-parser#230] [Q&A]
Hi @mliebelt,
Been trying to understand the Grammar file you had created for chess for a deeper understanding. Would be great if you could share some resources regarding **Parsing Expression Grammar (PEG)**... 😅

Would appreciate it a lot if others can share resources as well.... as I'm unable to find ample documentations and examples for the same. Apart from the **Peggy Documentation**

**Comment by mliebelt** [comment]
Thank you a lot for asking, even if I don't understand the question really. What exactly is on interest to you:

* Understanding how Peggy works? I have no idea, for me it is a black box.
* Understanding why some of the rules are way to complicated? Given the concrete rule, I may be able to explain why I have chosen that one, remembering different tries I did.
* Understanding how the JSON is constructed at the end? That was not so easy, I had to dig through a lot of resources myself, to find examples that looked similar and were easy enough to understand.

So I will collect some of the resources (I have used), that are not directly obvious when you look at Peggy. If you can be more specific with your question, that would help a lot. The grammar has been grown over time, and each change had a reason for it. [This gives you a view on the changes](https://github.com/mliebelt/pgn-parser/blame/main/src/pgn-rules.pegjs) of the grammar file. So depending which part is interesting, I could dig a little bit deeper into it.

Btw, should that discussion not moved over to pgn-parser instead? The PgnViewerJS is just a client using the pgn-parser, which defines the grammar.

**Comment by coffeeDev98** [comment]
I wanted to understand the syntax of the **.pegjs** files and some more examples to understand the different rules and usecases.
For example, the below image shows the only example in the documentation and a highlighted statement that mentions that the parsing begins at the **start** rule but, I was unable to find a similar rule in the grammar file you had created.

<img width="687" alt="image" src="https://user-images.githubusercontent.com/63295908/172861261-274ba678-df64-4834-8034-f5c64ac0cf6f.png">

Saying this, I  wanted to understand:

- Understanding why some of the rules are way to complicated? Given the concrete rule, I may be able to explain why I have chosen that one, remembering different tries I did.
- Understanding how the JSON is constructed at the end? That was not so easy, I had to dig through a lot of resources myself, to find examples that looked similar and were easy enough to understand.
- How to create a grammar file from scratch with rules and functions.


Also, yes, I shall move this discussion to **pgn-parser**

**Reply by mliebelt** [reply]
Ok, I will give it a try (and I am pretty sure that more questions will follow):

* Start rule: 
  * The default is, that the first rule (which is in my case `games`) is the default rule. This was added (if I remember correctly, `git blame` will show that) later, I started with one game first. I even think I started without tags, so the first default start rule was `pgn` (see [commit](https://github.com/mliebelt/pgn-parser/commit/b293144cb6208a01395ecf7b2ea6056498223e91#diff-e23b97b5abbe8d4534f8a0412f009843bb46683e12be2a74a1864ad360c345b2)).
  * Later on, I added other start rules: `game`, `tags` and `games` at the end. For that reason, the function `parse` has the `PgnOptions` as parameter, which contains the `StartRule` then.
* Some of the rules "way too complicated": For each of the rules, there was a reason to do it that way. There are many reasons:
  * the [pgn specification](https://github.com/mliebelt/pgn-spec-commented/blob/main/pgn-specification.md) demands it. When I look at other grammars (which are often much simpler), I can easily construct a pgn string, that is allowed by the spec, but not read or interpreted, or vice versae. My goal was all the time to have a _complete implementation_  of the spec.
  * PEG parsers have their problems. Left recursion is one of them, and sometimes you have to spread things over more rules, check beginnings, ... to avoid endless loops. And no, there is no documentation available, why a solution was chosen.
  * The whole tags part of the grammar is lengthy, but complete. And it includes many extensions that are used, but that are not part of the spec. You can do that much easier (just `tag = bl tagKeyValue br; tagKeyValue = string ws string`), but this allows many flaws, and will not help checking correctness of tags.
  * The `pgn` part is in my opinion as sparse as possible, but not sparser. There is one compromise done, that is reasonable: The spec allows comments everywhere, which is just not reasonable.
  * Half of the grammar comes from [PGN Spec extensions](https://github.com/mliebelt/pgn-spec-commented/blob/main/pgn-spec-supplement.md) which is completely implemented as well. Similar goal as above, allow to read everything that is possible under the sun, _but well formed according to a spec_.
  * Parts are deviations done by the _big shops_: lichess extensions, something from ChessTempo if I remember correctly, wrong implementation of chess.com of some small part. But I should look that up, to be more concrete.
* JSON construction: well, that is base Javascript, and if you don't want to have many functions, you have to construct the JSON by hand. If you have an idea, how to make that easier to read, file an issue about it, or make a pull request, that is a real improvement (and does not throw away any functionality that is now working).
* Construct a new grammar file: Well, the `pgn-parser` is no tutorial how to create a PEG parser, and it is not so easy to learn that from it. But I think there are many tutorials about PEG contruction available, which have exactly the deepness of your screenshot: a small mathematical grammar ... I had the same problem, and perhaps I can (re-)find the examples I have used in the construction. Don't expect too much from it.

I hope those comments help a little bit, and I try to be more helpful then if needed.

**Reply by coffeeDev98** [reply]
This gives me a good idea of where to start from now. Thanks a lot and really appreciate the response!! :)
I shall go through the PGN Specification shared above. Guess this is going to take some amount of time...

As for the PEG Parser resources, if you're able to find those examples again and if it's not much trouble, it'll help me and others in my situation as well (I think).

**Reply by mliebelt** [reply]
Ok, after searching now for some hours, I have not really found any useful additional resources. The ones that my help best are:

* https://github.com/peggyjs/peggy/tree/main/examples The `examples` directory in the original repository. Some of them stemming from the past (from `peg.js`), but some more new. I am pretty sure I have used them, and at least I found some usefule patterns like that head/tail usage for recursive things (see [link](https://github.com/peggyjs/peggy/blob/244029be7d8a693ef0b1dd90fc4c14ace2bba4ea/examples/javascript.pegjs#L126) for a concrete pattern)
* The documentation is very good, and I have read the part [about the grammar](https://peggyjs.org/documentation.html#grammar-syntax-and-semantics) numerous times to find gems in it. Especially the `& expression` and `! expression` were very helpful, and are usedin the grammar.
* Sometimes overlooked, the [wiki grammar page](https://github.com/peggyjs/peggy/wiki/Useful-Grammars) has some references, one going to an outside repository with a very ambitous goal (and very advanced grammars). Not easy stuff, but perhaps helpful. That outside usage has no result preparation included, by purpose, which makes it not so easy to use.
* https://groups.google.com/g/pegjs: Just to mention, there is an old google group, with mostly nonsense in it, but perhaps you are lucky and find there some resource not available elsewhere. I think it is mostly dead, but who knows.

So, sorry, no really useful references. But that means, you didn´t do anything wrong when searching for it ...

**Reply by coffeeDev98** [reply]
This is a lot of good info and resources. Thank you so much for taking the time and sending these resources. Appreciate it a lot! :)
